(venv) ➜  code git:(master) ✗ python3 -m pdb charges_calc.py --input=source.json --output=out.json
> /Users/i532701/UW_class/Python220/students/mgummel/lesson02/assignment/code/charges_calc.py(3)<module>()
-> '''
(Pdb) n
> /Users/i532701/UW_class/Python220/students/mgummel/lesson02/assignment/code/charges_calc.py(4)<module>()
-> import argparse
(Pdb) n
> /Users/i532701/UW_class/Python220/students/mgummel/lesson02/assignment/code/charges_calc.py(5)<module>()
-> import json
(Pdb) n
> /Users/i532701/UW_class/Python220/students/mgummel/lesson02/assignment/code/charges_calc.py(6)<module>()
-> import datetime
(Pdb) n
> /Users/i532701/UW_class/Python220/students/mgummel/lesson02/assignment/code/charges_calc.py(7)<module>()
-> import math
(Pdb) n
> /Users/i532701/UW_class/Python220/students/mgummel/lesson02/assignment/code/charges_calc.py(10)<module>()
-> def parse_cmd_arguments():
(Pdb) n
> /Users/i532701/UW_class/Python220/students/mgummel/lesson02/assignment/code/charges_calc.py(18)<module>()
-> def load_rentals_file(filename):
(Pdb) n
> /Users/i532701/UW_class/Python220/students/mgummel/lesson02/assignment/code/charges_calc.py(26)<module>()
-> def calculate_additional_fields(data):
(Pdb) n
> /Users/i532701/UW_class/Python220/students/mgummel/lesson02/assignment/code/charges_calc.py(40)<module>()
-> def save_to_json(filename, data):
(Pdb) n
> /Users/i532701/UW_class/Python220/students/mgummel/lesson02/assignment/code/charges_calc.py(44)<module>()
-> if __name__ == "__main__":
(Pdb) n
> /Users/i532701/UW_class/Python220/students/mgummel/lesson02/assignment/code/charges_calc.py(45)<module>()
-> args = parse_cmd_arguments()
(Pdb) s
--Call--
> /Users/i532701/UW_class/Python220/students/mgummel/lesson02/assignment/code/charges_calc.py(10)parse_cmd_arguments()
-> def parse_cmd_arguments():
(Pdb) n
> /Users/i532701/UW_class/Python220/students/mgummel/lesson02/assignment/code/charges_calc.py(11)parse_cmd_arguments()
-> parser = argparse.ArgumentParser(description='Process some integers.')
(Pdb) n
> /Users/i532701/UW_class/Python220/students/mgummel/lesson02/assignment/code/charges_calc.py(12)parse_cmd_arguments()
-> parser.add_argument('-i', '--input', help='input JSON file', required=True)
(Pdb) n
> /Users/i532701/UW_class/Python220/students/mgummel/lesson02/assignment/code/charges_calc.py(13)parse_cmd_arguments()
-> parser.add_argument('-o', '--output', help='ouput JSON file', required=True)
(Pdb) n
> /Users/i532701/UW_class/Python220/students/mgummel/lesson02/assignment/code/charges_calc.py(15)parse_cmd_arguments()
-> return parser.parse_args()
(Pdb) parser.parse_args()
Namespace(input='source.json', output='out.json')
(Pdb) n
--Return--
> /Users/i532701/UW_class/Python220/students/mgummel/lesson02/assignment/code/charges_calc.py(15)parse_cmd_arguments()->Namespace(inp...ut='out.json')
-> return parser.parse_args()
(Pdb) n
> /Users/i532701/UW_class/Python220/students/mgummel/lesson02/assignment/code/charges_calc.py(46)<module>()
-> data = load_rentals_file(args.input)
(Pdb) s
--Call--
> /Users/i532701/UW_class/Python220/students/mgummel/lesson02/assignment/code/charges_calc.py(18)load_rentals_file()
-> def load_rentals_file(filename):
(Pdb) ll
 18  -> def load_rentals_file(filename):
 19         with open(filename) as file:
 20             try:
 21                 data = json.load(file)
 22             except:
 23                 exit(0)
 24         return data
(Pdb) n
> /Users/i532701/UW_class/Python220/students/mgummel/lesson02/assignment/code/charges_calc.py(19)load_rentals_file()
-> with open(filename) as file:
(Pdb) n
> /Users/i532701/UW_class/Python220/students/mgummel/lesson02/assignment/code/charges_calc.py(20)load_rentals_file()
-> try:
(Pdb) filename
'source.json'
(Pdb) n
> /Users/i532701/UW_class/Python220/students/mgummel/lesson02/assignment/code/charges_calc.py(21)load_rentals_file()
-> data = json.load(file)
(Pdb) file
<_io.TextIOWrapper name='source.json' mode='r' encoding='UTF-8'>
(Pdb) n
> /Users/i532701/UW_class/Python220/students/mgummel/lesson02/assignment/code/charges_calc.py(24)load_rentals_file()
-> return data
(Pdb) n
--Return--
> /Users/i532701/UW_class/Python220/students/mgummel/lesson02/assignment/code/charges_calc.py(24)load_rentals_file()->{'RNT001': {'price_per_day': 31, 'product_code': 'PRD80', 'rental_end': '3/22/17', 'rental_start': '6/12/17', ...}, 'RNT002': {'price_per_day': 16, 'product_code': 'PRD11', 'rental_end': '9/30/18', 'rental_start': '7/20/16', ...}, 'RNT003': {'price_per_day': 40, 'product_code': 'PRD22', 'rental_end': '6/4/17', 'rental_start': '2/1/16', ...}, 'RNT004': {'price_per_day': 40, 'product_code': 'PRD86', 'rental_end': '12/7/17', 'rental_start': '8/14/16', ...}, ...}
-> return data
(Pdb) n
> /Users/i532701/UW_class/Python220/students/mgummel/lesson02/assignment/code/charges_calc.py(47)<module>()
-> data = calculate_additional_fields(data)
(Pdb) s
--Call--
> /Users/i532701/UW_class/Python220/students/mgummel/lesson02/assignment/code/charges_calc.py(26)calculate_additional_fields()
-> def calculate_additional_fields(data):
(Pdb) ll
 26  -> def calculate_additional_fields(data):
 27         for value in data.values():
 28             try:
 29                 rental_start = datetime.datetime.strptime(value['rental_start'], '%m/%d/%y')
 30                 rental_end = datetime.datetime.strptime(value['rental_end'], '%m/%d/%y')
 31                 value['total_days'] = (rental_end - rental_start).days
 32                 value['total_price'] = value['total_days'] * value['price_per_day']
 33                 value['sqrt_total_price'] = math.sqrt(value['total_price'])
 34                 value['unit_cost'] = value['total_price'] / value['units_rented']
 35             except:
 36                 exit(0)
 37     
 38         return data
(Pdb) n
> /Users/i532701/UW_class/Python220/students/mgummel/lesson02/assignment/code/charges_calc.py(27)calculate_additional_fields()
-> for value in data.values():
(Pdb) n
> /Users/i532701/UW_class/Python220/students/mgummel/lesson02/assignment/code/charges_calc.py(28)calculate_additional_fields()
-> try:
(Pdb) value
{'product_code': 'PRD80', 'units_rented': 8, 'price_per_day': 31, 'rental_start': '6/12/17', 'rental_end': '3/22/17'}
(Pdb) n
> /Users/i532701/UW_class/Python220/students/mgummel/lesson02/assignment/code/charges_calc.py(29)calculate_additional_fields()
-> rental_start = datetime.datetime.strptime(value['rental_start'], '%m/%d/%y')
(Pdb) n
> /Users/i532701/UW_class/Python220/students/mgummel/lesson02/assignment/code/charges_calc.py(30)calculate_additional_fields()
-> rental_end = datetime.datetime.strptime(value['rental_end'], '%m/%d/%y')
(Pdb) rental_end
*** NameError: name 'rental_end' is not defined
(Pdb) rental_start
datetime.datetime(2017, 6, 12, 0, 0)
(Pdb) n
> /Users/i532701/UW_class/Python220/students/mgummel/lesson02/assignment/code/charges_calc.py(31)calculate_additional_fields()
-> value['total_days'] = (rental_end - rental_start).days
(Pdb) rental_end
datetime.datetime(2017, 3, 22, 0, 0)
(Pdb) n
> /Users/i532701/UW_class/Python220/students/mgummel/lesson02/assignment/code/charges_calc.py(32)calculate_additional_fields()
-> value['total_price'] = value['total_days'] * value['price_per_day']
(Pdb) n
> /Users/i532701/UW_class/Python220/students/mgummel/lesson02/assignment/code/charges_calc.py(33)calculate_additional_fields()
-> value['sqrt_total_price'] = math.sqrt(value['total_price'])
(Pdb) value['total_price']
-2542
(Pdb) n
ValueError: math domain error
> /Users/i532701/UW_class/Python220/students/mgummel/lesson02/assignment/code/charges_calc.py(33)calculate_additional_fields()
-> value['sqrt_total_price'] = math.sqrt(value['total_price'])
(Pdb) n
> /Users/i532701/UW_class/Python220/students/mgummel/lesson02/assignment/code/charges_calc.py(35)calculate_additional_fields()
-> except:
(Pdb) n
> /Users/i532701/UW_class/Python220/students/mgummel/lesson02/assignment/code/charges_calc.py(36)calculate_additional_fields()
-> exit(0)
(Pdb) n
SystemExit: 0
> /Users/i532701/UW_class/Python220/students/mgummel/lesson02/assignment/code/charges_calc.py(36)calculate_additional_fields()
-> exit(0)
(Pdb) Traceback (most recent call last):
  File "/Users/i532701/UW_class/Python220/students/mgummel/lesson02/assignment/code/charges_calc.py", line 33, in calculate_additional_fields
    value['sqrt_total_price'] = math.sqrt(value['total_price'])
ValueError: math domain error

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pdb.py", line 1699, in main
    pdb._runscript(mainpyfile)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pdb.py", line 1568, in _runscript
    self.run(statement)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/bdb.py", line 585, in run
    exec(cmd, globals, locals)
  File "<string>", line 1, in <module>
  File "/Users/i532701/UW_class/Python220/students/mgummel/lesson02/assignment/code/charges_calc.py", line 47, in <module>
    data = calculate_additional_fields(data)
  File "/Users/i532701/UW_class/Python220/students/mgummel/lesson02/assignment/code/charges_calc.py", line 36, in calculate_additional_fields
    exit(0)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/bdb.py", line 94, in trace_dispatch
    return self.dispatch_exception(frame, arg)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/bdb.py", line 173, in dispatch_exception
    self.user_exception(frame, arg)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pdb.py", line 310, in user_exception
    self.interaction(frame, exc_traceback)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pdb.py", line 350, in interaction
    self._cmdloop()
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pdb.py", line 319, in _cmdloop
    self.cmdloop()
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/cmd.py", line 126, in cmdloop
    line = input(self.prompt)
ValueError: I/O operation on closed file.
Uncaught exception. Entering post mortem debugging
Running 'cont' or 'step' will restart the program
> /Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/cmd.py(126)cmdloop()
-> line = input(self.prompt)
(Pdb) Traceback (most recent call last):
  File "/Users/i532701/UW_class/Python220/students/mgummel/lesson02/assignment/code/charges_calc.py", line 33, in calculate_additional_fields
    value['sqrt_total_price'] = math.sqrt(value['total_price'])
ValueError: math domain error

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pdb.py", line 1699, in main
    pdb._runscript(mainpyfile)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pdb.py", line 1568, in _runscript
    self.run(statement)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/bdb.py", line 585, in run
    exec(cmd, globals, locals)
  File "<string>", line 1, in <module>
  File "/Users/i532701/UW_class/Python220/students/mgummel/lesson02/assignment/code/charges_calc.py", line 47, in <module>
    data = calculate_additional_fields(data)
  File "/Users/i532701/UW_class/Python220/students/mgummel/lesson02/assignment/code/charges_calc.py", line 36, in calculate_additional_fields
    exit(0)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/bdb.py", line 94, in trace_dispatch
    return self.dispatch_exception(frame, arg)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/bdb.py", line 173, in dispatch_exception
    self.user_exception(frame, arg)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pdb.py", line 310, in user_exception
    self.interaction(frame, exc_traceback)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pdb.py", line 350, in interaction
    self._cmdloop()
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pdb.py", line 319, in _cmdloop
    self.cmdloop()
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/cmd.py", line 126, in cmdloop
    line = input(self.prompt)
ValueError: I/O operation on closed file.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pdb.py", line 1726, in <module>
    pdb.main()
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pdb.py", line 1718, in main
    pdb.interaction(None, t)
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pdb.py", line 350, in interaction
    self._cmdloop()
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/pdb.py", line 319, in _cmdloop
    self.cmdloop()
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/cmd.py", line 126, in cmdloop
    line = input(self.prompt)
ValueError: I/O operation on closed file.
